brainiak/reprsimil/brsa.py

```diff
@@ -0,0 +1,1118 @@
+            logger.debug('initial fitting without GP prior'
+                         ' for {} times'.format(init_iter))
+
+            for it in range(0, init_iter):
+                res_fitV = scipy.optimize.minimize(
+                    self._loglike_y_AR1_diagV_fitV, param0_fitV[:n_V - 1],
+                    args=(XTX, XTDX, XTFX, YTY_diag, YTDY_diag, YTFY_diag,
+                          XTY, XTDY, XTFY, current_vec_U_chlsk_l_AR1,
+                          current_a1, l_idx, n_C, n_T, n_V, rank,
+                          False, False),
+                    method='BFGS', jac=True, tol=10**-2,
+                    options={'xtol': 10**-3, 'disp': self.verbose,
+                             'maxiter': 4})
+
+                current_logSNR2[0:n_V - 1] = res_fitV.x[0:n_V - 1]
+                current_logSNR2[-1] = - np.sum(current_logSNR2[0:n_V - 1])
+
+                norm_fitVchange = np.linalg.norm(res_fitV.x[:n_V - 1]
+                                                 - param0_fitV[:n_V - 1])
+                logger.debug('norm of parameter change after fitting V: '
+                             '{}'.format(norm_fitVchange))
+                logger.debug('E[log(SNR2)^2]:', np.mean(current_logSNR2**2))
+
+                # Debugging purpose. Commented out due to complexity limit.
+                # if np.any(np.logical_not(np.isfinite(current_logSigma2))):
+                #     logger.debug('Initial fitting: iteration {}'.format(it))
+                #     logger.debug('current log(sigma^2): '
+                #                  '{}'.format(current_logSigma2))
+                #     logger.debug('current log(SNR^2): '
+                #                  '{}'.format(current_logSNR2))
+                #     logger.debug('log(sigma^2) has non-finite number')
+
+                param0_fitV[:n_V - 1] = res_fitV.x.copy()
+
+                # fit U, the covariance matrix, together with AR(1) param
+                param0_fitU[0:n_l] = current_vec_U_chlsk_l_AR1
+                param0_fitU[n_l:n_l + n_V] = current_a1
+                res_fitU = scipy.optimize.minimize(
+                    self._loglike_y_AR1_diagV_fitU, param0_fitU,
+                    args=(XTX, XTDX, XTFX, YTY_diag, YTDY_diag, YTFY_diag,
+                          XTY, XTDY, XTFY, current_logSNR2, l_idx, n_C,
+                          n_T, n_V, rank),
+                    method='BFGS', jac=True, tol=10**-2,
+                    options={'xtol': 10**-3, 'disp': self.verbose,
+                             'maxiter': 3})
+                current_vec_U_chlsk_l_AR1 = res_fitU.x[0:n_l]
+                current_a1 = res_fitU.x[n_l:n_l + n_V]
+                norm_fitUchange = np.linalg.norm(res_fitU.x - param0_fitU)
+                logger.debug('norm of parameter change after fitting U: '
+                             '{}'.format(norm_fitUchange))
+                param0_fitU = res_fitU.x.copy()
+
+                # if norm_fitVchange / np.sqrt(param0_fitV.size - n_smooth)\
```

Comment:
Uncommenting out this convergence check will probably speed things up by not overdoing the optimization?